### ISSUE TITLE
Wide character in print

### ISSUE THREAD

**Issue by c-alpha** [debug-ito/Gnuplot-Builder#3]
Hello Toshio-san,

I got an error:
```
Wide character in print at /Users/alexanderadolf/Library/Perl/perl5/perls/perl-5.38.0/lib/site_perl/5.38.0/Gnuplot/Builder/Process.pm line 152.
```

After looking at `Process.pm`, and the documentation at https://metacpan.org/pod/Gnuplot::Builder, I concluded that I should add
```perl
$Gnuplot::Builder::Process::ENCODING = 'utf8';
```
to my script. And that fixed the issue for me (no more error message).

One way to help other users who might encounter this, could be to describe the encoding variable in the documentation at https://metacpan.org/pod/Gnuplot::Builder.

On the other hand, I am doing something like this in my script:
```perl
my $builder = Gnuplot::Builder::Script->new();
$builder->set(
  encoding => 'utf8',
  ...
);
```
Perhaps there could be a way for `Gnuplot::Builder::Script` to detect when the `encoding` option is set, and use that value for `$Gnuplot::Builder::Process::ENCODING`?

Many thanks, and looking forward to your thoughts.

**Comment by c-alpha** [comment]
Toshio Ito ***@***.***> writes:

> Thanks for raising this issue (and the other one, too). I'll look into
> them. It may take some time because I have been away from this project
> for a long time.

Many thanks for your swift response, and no worries about timing. As I
have a workaround, there's no urgency.

**Comment by debug-ito** [comment]
Luckily, I had some time today :)

> One way to help other users who might encounter this, could be to describe the encoding variable in the documentation

You are right. The `$ENCODING` variable is so hard to find, while this problem can be pretty common. I'll update the documentation.

> Perhaps there could be a way for Gnuplot::Builder::Script to detect when the encoding option is set, and use that value for $Gnuplot::Builder::Process::ENCODING?

As far as I read the document, there is no such option for `Gnuplot::Builder::Script`. However, I don't think it's necessary to configure character encoding on per-script basis. I think the global `$ENCODING` variable is sufficient.

If you really need the per-script encoding option, could you provide me some use case for that?

**Comment by c-alpha** [comment]
Apologies for the delay in getting back to you!

Toshio Ito ***@***.***> writes:

> [...]
> You are right. The `$ENCODING` variable is so hard to find, while this
> problem can be pretty common. I'll update the documentation.

Cool, thanks for taking this in board.

>> Perhaps there could be a way for Gnuplot::Builder::Script to detect
>> when the encoding option is set, and use that value for
>> $Gnuplot::Builder::Process::ENCODING?
>
> As far as I read the document, there is no such option for
> `Gnuplot::Builder::Script`. However, I don't think it's necessary to
> configure character encoding on per-script basis. I think the global
> `$ENCODING` variable is sufficient.
> [...]

After contemplating a bit over this, I agree with you; it is a global
setting.

Contemplating about it some more, I think I would still like to suggest
some kind of "adaptive behaviour". For instance, if I do
```perl
use open qw(:std :encoding(UTF-8));
```
near the top of my script, then from that point onwards
```perl
my @layers = PerlIO::get_layers(STDIN);
```
will return something like this:
```
[
    [0] "unix",
    [1] "perlio",
    [2] "encoding(utf-8-strict)",
    [3] "utf8"
]
```

The thing is, for as long as UTF-8 is not the global default for
everything, it can be quite a pain to go through all the modules you're
using, and (a) figuring out what each module's default encoding is, and
(b) figuring out how to set the desired encoding for each module.

In the spirit of promoting UTF-8, and of reducing the research burden on
module users, my - admittedly very personal - take would be to say
something along the lines of "make UTF-8 the default, and provide a
means for users to explicitly switch to any other encoding if they
must."

**Comment by debug-ito** [comment]
Sorry, I cannot understand the meaning of "adaptive behaviour". Could you explain your desired API (of Gnuplot::Builder) more explicitly?

> something along the lines of "make UTF-8 the default

I do understand this point, but I'm reluctant to make UTF-8 encoding the default. That's because (1) it would be a breaking change, and (2) Gnuplot::Builder is promoted as a thin layer.

**Comment by c-alpha** [comment]
Toshio Ito ***@***.***> writes:

> Sorry, I cannot understand the meaning of "adaptive behaviour". Could
> you explain your desired API (of Gnuplot::Builder) more explicitly?
> [...]

For instance, by using `PerlIO::get_layers(STDOUT)` you could determine
whether the user has set `:encoding(UTF-8)` on `STDOUT`. If and when
this is the case, `Gnuplot::Builder` could automatically use UTF-8
encoding when producing output, just as if the user had set `$ENCODING`
explicitly (but which would not be necessary in this case).

Thus, "adaptive behaviour" was intended to mean that `Gnuplot::Builder`
could behave differently, depending on other, global settings the user
has made, and which document the user's intent.

Hope this clarifies?

**Comment by debug-ito** [comment]
I see. Thanks for explanation. However, I'm not going to adopt that kind of adaptive behavior as the default due to the same two reasons in [the previous comment](https://github.com/debug-ito/Gnuplot-Builder/issues/3#issuecomment-1745150607).

Why not implement the adaptive encoding config as a separate module and publish it?

**Comment by c-alpha** [comment]
Toshio Ito ***@***.***> writes:

> [...]
> Why not implement the adaptive encoding config as a separate module
> and publish it?
> [...]

It already exists in the form of PerlIO
(https://perldoc.perl.org/PerlIO). See also `use open`
(https://perldoc.perl.org/open).

My "default preamble" for Perl code looks like this:
```perl
use strict;
use warnings;
use utf8;
use open qw(:std :encoding(UTF-8));
```

`use utf8` tells Perl that the script itself is encoded in UTF-8, since
I want to use things like "dBμV" (Decibel Microvolt) as part of strings
to print for instance.

`use open qw(:std :encoding(UTF-8))` tells perl to push the UTF-8
encoding layer onto STDIN, STDOUT, and STDERR.

As an end user, I think it would be great if any modules I am using
could use the same settings as STDIN for reading and as STDOUT for
writing, unless there is a very specific requirement that the data to be
read or written must be in a specific format. Such a "very specific
reason" could for example be that the file format in which the data is
to be read or written prescribes a specific encoding.

That said, and quickly skimming through the code of IPC::Open3, it would
seem as if using UTF-8 encoding with Gnuplot::Builder could perhaps
already work if the script had a `use open qw(:std :encoding(UTF-8))`
pragma, and line 150 of Gnuplot/Builder/Process.pm were commented out.
Perhaps you could do a little experiment to confirm whether this is the
case, or not?

If it did work that way, it would allow you to drop `$ENCODING`
altogether. Thus, instead of forcing the user to set an extra
environment, or Perl variable for this single purpose, the communication
channel with Gnuplot would rely on Perl's IO layer and inherit whatever
the user has configured for STDIN, STDOUT, and STDERR there. Less code
for you, less code and less confusion for end users. But it all depends
on whether IPC::Open3 is a "good Perl IO citizen."

**Comment by debug-ito** [comment]
As I already said,  I'm not going to adopt that kind of adaptive behavior as the default due to the same two reasons in [the previous comment](https://github.com/debug-ito/Gnuplot-Builder/issues/3#issuecomment-1745150607).

> It already exists in the form of PerlIO

I mean, you could make your own module, say Gnuplot::Builder::AdaptiveEncoding. When loaded, that module would read the encoding settings from PerlIO and set it to `$ENCODING`. I think that is your desired behavior. Publishing and promoting it would help developers that have similar opinions to you.

>  I think it would be great if any modules I am using could use the same settings as STDIN for reading and as STDOUT for writing

I disagree with you at this point. For me, character encoding should be handled by the application. The application should be aware of what kind of encoding and data format it should use for every filehandle, unless it's hidden deeply inside middlewares. So, I think adopting configuration of STDIN for every other reading channel would violate the responsibility boundary of a module. That might break the application's assumption on those channels.

Of course, I wouldn't stop you making and publishing a module that does the above job.

**Comment by c-alpha** [comment]
Toshio Ito ***@***.***> writes:

> [...]
> I disagree with you at this point. For me, character encoding should
> be handled by the application. The application should be aware of what
> kind of encoding and data format it should use for every filehandle,
> unless it's hidden deeply inside middlewares. So, I think adopting
> configuration of STDIN for every other reading channel would violate
> the responsibility boundary of a module. That might break the
> application's assumption on those channels.
> [...]

I think I was trying to express something very similar as you are saying
here. Which probably means I failed to explain myself well enough.

Yes, of course the application must be able to exercise full control
over the encodings used; if and when it wants to.

But what if the application does not want to explicitly control it? As
we both agreed before, the encoding to use is not per script, but rather
a global setting. So from an application developer's point of view, it
would be very convenient to be able to define a default encoding, and
all in-/output would use that encoding unless a different encoding was
explicitly specified for that particular in-/output channel.

More research into this brought me to [1], [2], and [3] (where [1]
basically points at the other two). And as probably was to be expected,
the problem is much more complex than I thought.

[1] https://www.perlmonks.org/?node_id=1042907
[2] http://stackoverflow.com/a/6163129/827379
[3] http://stackoverflow.com/a/6192088/827379

In that light, I rest my case and am happy with `$ENCODING` being added
to the documentation.

Many thanks for your patience!
